### PR TITLE
Add FlyteNonRecoverableSystemException

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
NA

## Why are the changes needed?
Flyte returns the task when it fails to convert the Python output to a literal or vice versa.

## What changes were proposed in this pull request?
Raise `FlyteNonRecoverableSystemException` if flytekit fails to serialize the output. (e.g. failed to convert python.int to Literal.simple.string).
However, if the error comes from data persistence, we won't change the error type to `FlyteNonRecoverableSystemException`

## How was this patch tested?
remote cluster

```bash
pyflyte run --remote flyte-example/improve_error/mismatching_output.py wf
```

### Setup process
```python
from flytekit import task, workflow, ImageSpec

new_flytekit = "git+https://github.com/flyteorg/flytekit@3ef9b11cba76a237bb010c8a764ecdc2c2093089"

image_spec = ImageSpec(
    registry="pingsutw",
    apt_packages=["git"],
    packages=[new_flytekit],
    builder="envd",
)


@task(container_image=image_spec)
def t1() -> int:
    return "hello"


@workflow()
def wf():
    t1()
```

### Screenshots

Before:
Screenshot 2024-08-21 at 4 54 55 PM

After:
Screenshot 2024-08-21 at 4 54 47 PM

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
NA

## Docs link
NA

### CODE REVIEW

File: flytekit/bin/entrypoint.py [hunk 1]
```diff
@@ -158,14 +159,21 @@ def _dispatch_execute(
...
         logger.error("!! End Error Captured by Flyte !!")
 
-    # All the Non-user errors are captured here, and are considered system errors
+    # All the Non-user runtime errors are captured here, and are considered system errors
     except Exception as e:
-        exc_str = get_traceback_str(e)
+        kind = _error_models.ContainerError.Kind.RECOVERABLE
+        err = e
+
+        if isinstance(e, FlyteNonRecoverableSystemException):
```
Comment (thomasjpfan) [inline flytekit/bin/entrypoint.py:167]: Can this be another `except FlyteNonRecoverableSystemException as e` branch?

Comment (pingsutw) [reply flytekit/bin/entrypoint.py]: Good call. updated it

**Comment by wild-endeavor** [comment]
tried to merge in master, didn't do it correctly though, definitely missing some things, like the double `Exception` catching that's now in entrypoint.

Regardless I think we should think more about this - I'm not sure it really addresses the issue raised in [slack](https://flyte-org.slack.com/archives/CP2HDHKE1/p1724409985371659) (and commented in this [issue here](https://github.com/flyteorg/flyte/issues/381#issuecomment-2310297251)).  This issue is on the propeller side.

**Comment by wild-endeavor** [comment]
forgot to hit submit on that comment.  that was from before kevin fixed the merge

**Review by wild-endeavor — APPROVED** [review]

**Pull request merged**